Attempt at multiarch support (#545)

scpeters opened the pull request:
This is an attempt at multiarch support (#545), supercedes
(#604). The dpkg-architecture command is used to identify
an appropriate lib folder suffix to use for LD_LIBRARY_PATH
and PKG_CONFIG_PATH. builder.py and
catkin_generate_environment.cmake are modified
accordingly, ~~though there's still a problem with isolated
devel spaces~~.

~~It works for `catkin_make_isolated --install`, but I'm getting the following error when running `catkin_make_isolated`:~~

@wjwwood

dirk-thomas commented:
I will add some comments inline.

But the PR seems to fail unit tests anyway and as you mentioned yourself does not yet work in all cases.

dirk-thomas commented on cmake/catkin_generate_environment.cmake:
```diff
@@ -10,6 +10,20 @@ function(catkin_generate_environment)
+    )
+    if (NOT "${CATKIN_MULTIARCH}" STREQUAL "")
+      set(CATKIN_MULTIARCH_LIB_DESTINATION "['lib', 'lib/${CATKIN_MULTIARCH}']")
+      set(CATKIN_MULTIARCH_PKGCONFIG_DESTINATION "['lib/pkgconfig', 'lib/${CATKIN_MULTIARCH}/pkgconfig']")
```
These two lines should not repeat the values from lines 14 and 15 but use the variable to avoid duplication. Also there is a variable `CATKIN_GLOBAL_LIB_DESTINATION` which could be uses instead of using `lib`.

wjwwood commented:
> But the PR seems to fail unit tests anyway and as you mentioned yourself does not yet work in all cases.

I asked him to open it so I could help him with it.

dirk-thomas commented on cmake/templates/_setup_util.py.in:
```diff
@@ -151,11 +154,16 @@ def _prefix_env_variable(environ, name, paths, subfolder):
+        if type(subfolders) != type([]):
+            subfolders = [subfolders]
+        for subfolder in subfolders:
+            path = path0
```
The way `path` is stored in another variable and then overwritten in the loop looks pretty weird. Please avoid that by just using a different variable inside the loop to build the full path without changing the surrounding loop variable `path`.

scpeters replied:
Thanks, I think I fixed it in 232d56b

dirk-thomas commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,10 @@ def isolation_print_command(cmd, path=None, add_env=None):
+def get_multiarch(build_dir):
+    # this function returns the suffix for lib directories on supported systems or an empty string
+    return run_command('dpkg-architecture -qDEB_HOST_MULTIARCH 2> /dev/null', build_dir, quiet=True)
+
```
Please add an extra empty line to comply with PEP8.

dirk-thomas replied:
This code block does not look like it would work on Windows or even systems which don't have `dpkg-architecture` available.

wjwwood replied:
Yeah, you should either check for it's existence in the PATH first or catch an `OSError` when it is not found.

scpeters replied:
I added an extra line for PEP8 in 948eb7d.

wjwwood commented:
If you are using Python to invoke the `dpkg-architecture` command, then there is no need for the `2> /dev/null`. I would invoke it like this in python:

``` python
import subprocess
...
out, err = subprocess.Popen(
    ['dpkg-architecture', '-qDEB_HOST_MULTIARCH'],
    stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True).communicate()
# Use the returned data in `out`
```

scpeters commented:
The test_setup_util.py test is failing because I've changed the variables that are used to configure `_setup_util.py`. I've tried modifying this, but I'm having trouble debugging, since it writes to a temporary file that I can't see.

scpeters commented:
I fixed the test, and I've responded to all the comments except the one indicating that I should check for the existence of `dpkg-architecture`. I agree; though I'm able to run this code on my mac without any problem, even though it doesn't have `dpkg-architecture`.

scpeters commented:
I need to take a little break from this, if someone else is willing to help tackle checking the existence of `dpkg-architecture`. @j-rivero

j-rivero commented:
`dpkg-architecture` is part of `dpkg-dev` package which [seems not be part of a clean ubuntu installation](http://releases.ubuntu.com/releases/14.04/ubuntu-14.04-desktop-amd64.manifest).

Maybe we could use the debian/ubuntu print-multiarch gcc flag (present in all current Ubuntu distributions): `gcc -print-multiarch` which I believe should be present in all default installations out there.

scpeters commented:
`dpkg-dev` may not be part of a clean ubuntu installation, but it is [included with build-essential](http://packages.ubuntu.com/trusty/build-essential). How much can you do with catkin without `build-essential`?

j-rivero commented:
A [quick patch](https://github.com/j-rivero/catkin/commit/dcbc8f34adb698d5f4dd35d291126f98a802dc68) against Steve's branch seems to work fine in my system.

**_Error cases:**_ 
- CMake check for `gcc -print-multiarch` won't display anything by the output (ERROR_QUIET) and return empty if no gcc is present or no `-print-multiarch` flag exists in gcc.
- Same for python subprocess call, return an empty in cases of error.

scpeters commented:
Thanks @j-rivero ; I will defer judgment to the catkin maintainers about the patch.

j-rivero commented:
We can go with the approach of using dpkg-architecture, no a problem. Let me know if that is the option chosen and I will implement the check for the binary.

scpeters commented:
@wjwwood @dirk-thomas Do you have a preference between `dpkg-architecture` and `gcc -print-multiarch` (see Jose's comments above)?

I'll cc @thomas-moulard as well.

wjwwood commented:
First off, if `dpkg-architecture` is not available it should never fail. As long as that is the case I'm fine.

If `dpkg-architecture` is part of `build-essential` then it will be brought in as part of the base system for building ROS and it shouldn't be a problem. If you believe `gcc -print-multiarch` is more portable or reliable then I don't mind either one.

dirk-thomas commented:
Agreeing to Williams comment. Any tool which is either available by default on Debian systems or is being pulled in by the base dependencies of ROS is fine with me. Important is to consider custom configurations: can a user use clang and not have gcc? Or can he remove dpkg-architecture while still having everything else to build ROS packages?

scpeters commented:
As of trusty, [build-essential](http://packages.ubuntu.com/trusty/build-essential) depends on both `dpkg-dev` and `gcc`. An unrealistic edge case: if someone built clang from source without `build-essential` installed, then neither method would be available.

I personally would lean towards `dpkg-architecture`.

dirk-thomas commented:
Sounds good to me.

j-rivero commented:
I'm without enough caffeine but, what about using a two step approach? First check for `gcc -print-multiarch` and, if failed, try to use `dpkg-architecture` so we cover all the possible options.

I have implemented this, the [full patch would look like this](https://github.com/j-rivero/catkin/compare/multiarch_gcc). The [cmake change](https://github.com/j-rivero/catkin/compare/multiarch_gcc#diff-192246ff4741052b709c2a2e17fa27deR13) and the [python get_multiarch](https://github.com/j-rivero/catkin/compare/multiarch_gcc#diff-616a0c8f980657f6b84468d0ce4abc77R251). And a [lovely assert](https://github.com/j-rivero/catkin/compare/multiarch_gcc#diff-616a0c8f980657f6b84468d0ce4abc77R264).

scpeters commented:
@j-rivero I just added your patch. I think it should be enough to cover the case in which those binaries aren't installed, since it should just complain to `std_err`; I don't think subprocess will throw an exception. The std_out should be empty, and your assertion will verify that it matches the triplet pattern.

j-rivero commented:
@scpeters in my experiments, when declaring `shell=True` there is no exception if the binaries are not present.

scpeters commented:
I just squashed them all into scpeters@35d9827

j-rivero commented:
Thanks @scpeters. Let me know if there is anything else missing for merging.

scpeters commented:
I think it's ready to review.

wjwwood commented:
I'll try to review it today.

wjwwood commented on cmake/catkin_generate_environment.cmake:
```diff
@@ -10,6 +10,30 @@ function(catkin_generate_environment)
   endif()
 
+  # get multiarch name
+  set(CATKIN_MULTIARCH_LIB_DESTINATION "'${CATKIN_GLOBAL_LIB_DESTINATION}'")
```
From a purely semantic point of view, the name `CATKIN_MULTIARCH_LIB_DESTINATION` isn't great because it always contains a non arch-specific path (`CATKIN_GLOBAL_LIB_DESTINATION`) and can have multiple destinations in it. What about replacing `CATKIN_MULTIARCH_LIB_DESTINATION` with `CATKIN_LIB_ENVIRONMENT_PATHS` or something like that which is more descriptive of the contents and plural.

scpeters replied:
Thanks, I changed to `CATKIN_*_ENVIRONMENT_PATHS` in scpeters@655012a

wjwwood commented on cmake/catkin_generate_environment.cmake:
```diff
@@ -10,6 +10,30 @@ function(catkin_generate_environment)
 
+  # get multiarch name
+  set(CATKIN_MULTIARCH_LIB_DESTINATION "'${CATKIN_GLOBAL_LIB_DESTINATION}'")
+  set(CATKIN_MULTIARCH_PKGCONFIG_DESTINATION "os.path.join('${CATKIN_GLOBAL_LIB_DESTINATION}', 'pkgconfig')")
```
Same thing applies to `CATKIN_MULTIARCH_PKGCONFIG_DESTINATION`.

wjwwood commented on cmake/templates/_setup_util.py.in:
```diff
@@ -68,11 +68,14 @@ def rollback_env_variables(environ, env_var_subfolders):
-            environ[key] = value
-            lines.append(assignment(key, value))
+        subfolders = env_var_subfolders[key]
+        if type(subfolders) != type([]):
```
Use `isinstance` instead:

``` python
if not isinstance(subfolders, list):
```

Or if you want to be a little more forgiving on the type:

``` python
import collections
...
if not isinstance(subfolders, collections.Iterable):
```

scpeters replied:
I switched to isinstance in scpeters@655012a. Also, I tinkered with `collections.Iterable`, but that doesn't work since strings are Iterable. So I just used your first suggestion.

wjwwood replied:
Ah yeah, I was thinking that wasn't the case, you'd have to do something like check for the `__iter__` attribute or something. This is fine.

wjwwood commented on cmake/templates/_setup_util.py.in:
```diff
@@ -151,11 +154,15 @@ def _prefix_env_variable(environ, name, paths, subfolder):
-        # exclude any path already in env and any path we already added
-        if path not in environ_paths and path not in checked_paths:
-            checked_paths.append(path)
+        if type(subfolders) != type([]):
```
Same thing here.

wjwwood commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,22 @@ def isolation_print_command(cmd, path=None, add_env=None):
+    out, err = subprocess.Popen(
+        ['gcc -print-multiarch'],
+        stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True).communicate()
+    if err:
```
This is a weak check, it could fail and not print anything to stderr, I would instead do this:

``` python
p = subprocess.Popen(...)
out, err = p.communicate()
if p.returncode != 0:
   ...
```

scpeters replied:
Thanks, updated in scpeters@655012a

wjwwood commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,22 @@ def isolation_print_command(cmd, path=None, add_env=None):
     )
 
 
+def get_multiarch():
```
This whole function could be predicated on `if sys.platform == 'linux':` which would save the subprocess calls on OS X and stuff like that.

scpeters replied:
I added this check in scpeters@a0eb2a0, with slightly different logic, since sys.platform resolves to `linux2` if using kernel 2.x, and `linux3` for kernel 3.x.

wjwwood replied:
Good catch, thanks.

wjwwood replied:
Oh man, and it resolves to `linux` in both cases with Python 3.3+ :confounded:

wjwwood replied:
I guess I actually used [platform.system()](https://docs.python.org/2/library/platform.html#platform.system) in the past.

wjwwood commented on test/unit_tests/test_setup_util.py:
```diff
@@ -7,7 +7,8 @@
                       {
-                          'CATKIN_GLOBAL_LIB_DESTINATION': 'lib',
+                          'CATKIN_MULTIARCH_LIB_DESTINATION': '\'lib\'',
+                          'CATKIN_MULTIARCH_PKGCONFIG_DESTINATION': 'os.path.join(\'lib\', \'pkgconfig\')',
```
What lead you to change this to have escaping?

scpeters replied:
One of the tests was failing, and this fixed it. I don't remember the exact details. I can try reverting it and observe the failures to jog my memory.

wjwwood commented:
Generally looks good, I just had a few comments and questions.

scpeters commented:
@j-rivero I inverted the logic in an if statement in catkin_generate_environment.cmake in https://github.com/scpeters/catkin/commit/655012a59b06825732ec6a31594f69b8670a4084 because it looked backwards to me. Can you confirm?

j-rivero commented:
Ouch, the change looks good to me. Thanks for fixing it.

scpeters commented on cmake/templates/_setup_util.py.in:
```diff
@@ -52,9 +52,9 @@ IS_WINDOWS = (system == 'Windows')
+    'LD_LIBRARY_PATH' if not IS_DARWIN else 'DYLD_LIBRARY_PATH': @CATKIN_LIB_ENVIRONMENT_PATHS@,
     'PATH': '@CATKIN_GLOBAL_BIN_DESTINATION@',
-    'PKG_CONFIG_PATH': 'lib/pkgconfig',
+    'PKG_CONFIG_PATH': @CATKIN_PKGCONFIG_ENVIRONMENT_PATHS@,
```
@wjwwood This is why I had to add the escaping to the test. Previous the `CATKIN_*` variables were strings enclosed in single quotes, but I wanted them to be defined as a list instead, so they are resolved without surrounding single quotes. In the non-multiarch case that the test uses, it has to add its own single quotes, which require escaping. Let me know if that makes no sense.

wjwwood replied:
Yeah that makes sense to me.

scpeters commented:
Ok, I think I've addressed all comments for now.

wjwwood commented:
Ok, looks ok to me, @dirk-thomas

dirk-thomas commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,25 @@ def isolation_print_command(cmd, path=None, add_env=None):
+        stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
+    out, err = p.communicate()
+    if p.returncode != 0:
+      out, err = subprocess.Popen(
```
Please fix indentation.

scpeters replied:
fixed in scpeters@e99145a

dirk-thomas commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,25 @@ def isolation_print_command(cmd, path=None, add_env=None):
+    out, err = p.communicate()
+    if p.returncode != 0:
+      out, err = subprocess.Popen(
+          ['dpkg-architecture -qDEB_HOST_MULTIARCH'],
```
Passing both parts in a single string looks wrong to me. I think it needs to be a list of two string.

wjwwood replied:
I think this is allowed because of `shell=True`.

tfoote replied:
You can use a string with whitespace if shell=True, It would be clearer to drop the list. If you have a multi-element list with whitespace it seems to drop the following elements in the list.

scpeters replied:
I just removed the `shell=True` and switched to using a list in scpeters@e99145a. In previous attempts, I had never tried without `shell=True`, and it only worked as a string with whitespace. It does work, however, with a list and `shell=False`.

dirk-thomas commented on test/unit_tests/test_setup_util.py:
```diff
@@ -7,7 +7,8 @@
 data = configure_file(os.path.join(os.path.dirname(__file__), '..', '..', 'cmake', 'templates', '_setup_util.py.in'),
                       {
-                          'CATKIN_GLOBAL_LIB_DESTINATION': 'lib',
+                          'CATKIN_LIB_ENVIRONMENT_PATHS': '\'lib\'',
```
Please use double quotes outside to avoid escaping inner quotes.

scpeters replied:
Done in scpeters@e99145a

dirk-thomas commented on test/unit_tests/test_setup_util.py:
```diff
@@ -7,7 +7,8 @@
                       {
-                          'CATKIN_GLOBAL_LIB_DESTINATION': 'lib',
+                          'CATKIN_LIB_ENVIRONMENT_PATHS': '\'lib\'',
+                          'CATKIN_PKGCONFIG_ENVIRONMENT_PATHS': 'os.path.join(\'lib\', \'pkgconfig\')',
```
Please use double quotes outside to avoid escaping inner quotes.

scpeters replied:
Done in scpeters@e99145a

dirk-thomas commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,25 @@ def isolation_print_command(cmd, path=None, add_env=None):
+    # it uses two step approach to look for multiarch: first run gcc -print-multiarch and if
+    # failed try to run dpkg-architecture
+    p = subprocess.Popen(
+        ['gcc -print-multiarch'],
```
Passing both parts in a single string looks wrong to me. I think it needs to be a list of two string.

wjwwood replied:
Same as the other one.

dirk-thomas commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,25 @@ def isolation_print_command(cmd, path=None, add_env=None):
+    # failed try to run dpkg-architecture
+    p = subprocess.Popen(
+        ['gcc -print-multiarch'],
+        stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
```
Please remove the shell keyword since it seems to be unnecessary.

dirk-thomas commented on python/catkin/builder.py:
```diff
@@ -248,6 +248,25 @@ def isolation_print_command(cmd, path=None, add_env=None):
+    if p.returncode != 0:
+      out, err = subprocess.Popen(
+          ['dpkg-architecture -qDEB_HOST_MULTIARCH'],
+          stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True).communicate()
```
Please remove the shell keyword since it seems to be unnecessary.

wjwwood replied:
Are you sure it is not needed? I was under the impression that they needed the shell env to make this work? If not then I agree, drop the `shell=True` and use a list for the arguments.

dirk-thomas replied:
I was able to run it in an interactive Python shell without the shell keyword and both commands returned the expected output.

scpeters replied:
I just removed the shell=True and switched to using a list in scpeters@e99145a. In previous attempts, I had never tried without shell=True, and it only worked as a string with whitespace. It does work, however, with a list and shell=False.

scpeters commented:
Ok, I think that's all the comments and travis is happy. Thanks for the reviews and let me know if anything else needs work.

dirk-thomas commented on cmake/catkin_generate_environment.cmake:
```diff
@@ -10,6 +10,30 @@ function(catkin_generate_environment)
+    endif()
+    if (NOT "${CATKIN_MULTIARCH}" STREQUAL "")
+      set(CATKIN_LIB_ENVIRONMENT_PATHS "['${CATKIN_GLOBAL_LIB_DESTINATION}', os.path.join('${CATKIN_GLOBAL_LIB_DESTINATION}', '${CATKIN_MULTIARCH}')]")
+      set(CATKIN_PKGCONFIG_ENVIRONMENT_PATHS "[os.path.join('${CATKIN_GLOBAL_LIB_DESTINATION}', 'pkgconfig'), os.path.join('${CATKIN_GLOBAL_LIB_DESTINATION}', '${CATKIN_MULTIARCH}', 'pkgconfig')]")
```
Please update these two lines to not repeat the same value again but extend the value like this:

```
set(foo "[${foo}, newstuff]")
```

scpeters replied:
I fixed it and then resquashed.

scpeters committed:
Implement multiarch support in catkin (#545)

This implements multiarch support (#545), and supercedes
(#604). A two step approach uses gcc -print-multiarch and,
if failed, fallback to use dpkg-architecture to identify
an appropriate lib folder suffix to use for LD_LIBRARY_PATH
and PKG_CONFIG_PATH. builder.py and
catkin_generate_environment.cmake are modified
accordingly.

Thanks to @j-rivero for his contributions, which are hidden in
this squashed commit.

dirk-thomas commented:
Thank you, Steve (and everybody else), for your work on this!

thomas-moulard commented:
Thank you! :+1:

dirk-thomas commented:
This PR broke `catkin_make_isolated`:

```
KeyError: 'CATKIN_LIB_ENVIRONMENT_PATHS'
KeyError: 'CATKIN_PKGCONFIG_ENVIRONMENT_PATHS'
```

I fixed it in f89f4e83d2d7cdf126c185424720b66f0ee0e937.

scpeters commented:
Ok, thanks for fixing it. Maybe a test should be added for invoking from sym-linked folders?

dirk-thomas commented:
Sorry, I linked the wrong ticket number - the fix was f89f4e83d2d7cdf126c185424720b66f0ee0e937. The command `catkin_make_isolated` did not work anymore.

j-rivero commented:
Sounds to me like we should define a regression test to avoid situations like this one, where we were not able to detect the problem before commit.

dirk-thomas commented:
I completely agree with having a unit test. But you should still run the tools affected by a change manually and not solely rely on the tests to cover it. In this case any invocation of cmi would have exposed the regression.

dirk-thomas commented:
Both cases covered now: abf3fcbd1769be146f6d6f29cf5f9da3a1838f06 and 284bee39d2f6ddb9b343367c43b8505a4c04c42e

j-rivero commented:
Nice! thanks @dirk-thomas
